### PR TITLE
Fixed #550 : Added necessary allignment between glyph creation and ac…

### LINKED ISSUE #550: Cylinder direction not unique.
## Description
Directions passed to the cylinder actor do not refer to the unique axis of the cylinder.

## Way to reproduce
from fury import actor,window
import numpy as np

scene = window.Scene()

centers = np.array([[-2,0,0],[0,0,0],[2,0,0]])
dirs = np.array([[1,0,0],[0,1,0],[0,0,1]])

arrows = actor.arrow(centers, dirs, dirs)
scene.add(arrows)

centers += np.array([0,1,0])
cyls = actor.cylinder(centers, dirs, dirs)
scene.add(cyls)

showm = window.ShowManager(scene)
showm.initialize()
showm.start()

### PR DESCRIPTION
**Pull request by Sassafrass6**
Added necessary alignment between glyph creation and actor mapping. If a vtkSource is not aligned with the X axis, provide 4x4 orientation matrix to repeat_sources to align.

This is necessary for the Cylinder object, which is aligned to the global Y axis in vtk. Without the correct orientation of the glyph, rotation into the z plane is not possible.

Also, the DiskSource is default aligned to the Z axis in vtk.
A similar routine for plotting discs (i.e. another use case for this patch) is included in actor.py.

Cylinder's with same specified direction as Arrows, original:
Screen Shot 2022-03-03 at 5 28 53 PM

Trivial implementation of Disks (Non-functional in a similar manner to cylinders):
Screen Shot 2022-03-03 at 5 35 00 PM

Correct implementation of Cylinders and Disks:
Screen Shot 2022-03-03 at 5 26 16 PM

### CODE REVIEW

**Comment by skoudoro** [comment]
Hi @Sassafrass6, 

Thank you for this PR. Quite a busy period for me, Not sure I will have time to look at your PR this week or next week. 

@guaje, Can you look at this, please?

**Comment by skoudoro** [comment]
Could you also address all pep8 issues as you can see [here](https://github.com/fury-gl/fury/pull/551#issuecomment-1058750569)

**Comment by Sassafrass6** [comment]
I have addressed all of the PEP8 issues. 

There is apparently still some issue with the cylinder normals. They are not shaded as one would expect.

**Review by guaje — CHANGES_REQUESTED** [review]
Hi @Sassafrass6

Thanks for catching and fixing this. I have added some specific comments and I would like to ask you to add a new function to test the new `orientation` parameter. You can add it to `fury.tests.test_utils.py`.

File: fury/actor.py
Comment (Sassafrass6) [reply fury/actor.py]: Corrected

File: fury/actor.py
Comment (Sassafrass6) [reply fury/actor.py]: Done

File: fury/actor.py
Comment (Sassafrass6) [reply fury/actor.py]: Fixed in both cylinders and disks

File: fury/actor.py
Comment (Sassafrass6) [reply fury/actor.py]: Moved orientation to a new line

File: fury/utils.py
Comment (Sassafrass6) [reply fury/utils.py]: Done

File: fury/tests/test_actors.py
Comment (Sassafrass6) [reply fury/tests/test_actors.py:954]: Testing the orientation argument is not possible without a test on repeat_sources. Also, I would argue against creating a separate test for the disk actor, since it is an advanced geometry actor.
I've added arguments to test the actor-specific parameters, though.

File: fury/actor.py
Comment (Sassafrass6) [reply fury/actor.py:1634]: Done

File: fury/tests/test_actors.py
Comment (guaje) [reply fury/tests/test_actors.py:954]: The purpose of the separate test is to remove the unnecessary ifs. Another way to have a single test that still looks clean is to do something like this:

```python
def test_advanced_geometry_actor(interactive=False):
    xyz = np.array([[0, 0, 0], [50, 0, 0], [100, 0, 0]])
    dirs = np.array([[0.5, 0.5, 0.5], [0.5, 0, 0.5], [0, 0.5, 0.5]])

    heights = np.array([5, 7, 10])

    actor_list = [[actor.cone, {'heights': heights, 'resolution': 8}],
                  [actor.arrow, {'heights': heights, 'resolution': 9}],
                  [actor.cylinder, {'heights': heights, 'resolution': 10}],
                  [actor.disk, {'rinner': 4, 'router': 8, 'rresolution': 2,
                                'cresolution': 2}]]

    scene = window.Scene()

    for act_func, extra_args in actor_list:
        colors = np.array([[1, 0, 0, 0.3], [0, 1, 0, 0.4], [1, 1, 0, 1]])

        geom_actor = act_func(xyz, dirs, colors, **extra_args)
        scene.add(geom_actor)

        if interactive:
            window.show(scene, order_transparent=True)
        arr = window.snapshot(scene)
        report = window.analyze_snapshot(arr, colors=colors)
        npt.assert_equal(report.objects, 3)

        scene.clear()

        colors = np.array([1.0, 1.0, 1.0, 1.0])

        geom_actor = act_func(xyz, dirs, colors, **extra_args)
        scene.add(geom_actor)

        if interactive:
            window.show(scene, order_transparent=True)
        arr = window.snapshot(scene)
        report = window.analyze_snapshot(arr, colors=[colors])
        npt.assert_equal(report.objects, 3)

        scene.clear()
```

File: fury/utils.py
Comment (guaje) [inline fury/utils.py:709]: > Testing the orientation argument is not possible without a test on repeat_sources.

Yes, please do add a test in `fury.tests.test_utils.py`

**Review by guaje — CHANGES_REQUESTED** [review]
@Sassafrass6 I have added some additional requests based on your latest changes/commits.

File: fury/tests/test_actors.py
Comment (Sassafrass6) [reply fury/tests/test_actors.py]: Agreed, looks nice.  I've updated the branch.

File: fury/utils.py
Comment (Sassafrass6) [reply fury/utils.py:709]: Your implementation of test_advanced_geometry_actor is nice, thank you. I've committed the change.

repeat_sources is used by several actors. Testing its functionality requires careful thought and definitely some effort. 
Because I did not author this routine, and am only trying to submit a bug fix to the code, I respectfully decline.

Comment (guaje) [reply fury/utils.py:709]: I understand. You have addressed all my comments. IDK if @skoudoro has any more comments.

**Comment by Sassafrass6** [comment]
Any update on the acceptance of this PR?

The modification is critical for the cylinder actor to function.

**Comment by skoudoro** [comment]
Hi @Sassafrass6,

Apologize for not giving news. The core maintainers (including me) are really focused on this workshop that we are organizing: https://dipy.org/workshops/dipy-workshop-2022 

That's why we were not so active these past 2 - 3 weeks. The workshop finishes this Friday, so everything will get back to normal, next week (review, update, etc...).

I agree this is an important update which might imply a quick-release also.

**Comment by skoudoro** [comment]
> This is necessary for the Cylinder object, which is aligned to the global Y axis in vtk

Should not it be applied to all actors? I am surprised that only the cylinders are oriented along the Y-axis.  I suppose this is the case for all vtksource.  Which vtksource is aligned along the X-axis? 

I need to look deeper at this, [when I see this tutorial](https://kitware.github.io/vtk-examples/site/Cxx/GeometricObjects/GeometricObjectsDemo/), they seem to have all the same orientations.

File: fury/actor.py
Comment (skoudoro) [inline fury/actor.py:1614]: You need to keep this. It allows using your own faces and vertices if you build yourself a cylinder.

it give us flexibility for future change in the API

File: fury/actor.py
Comment (skoudoro) [inline fury/actor.py:1646]: `float, optional` instead

File: fury/actor.py
Comment (skoudoro) [inline fury/actor.py:1648]: `float, optional` instead

File: fury/actor.py
Comment (skoudoro) [inline fury/actor.py:1650]: `int, optional` instead

File: fury/actor.py
Comment (skoudoro) [inline fury/actor.py:1652]: `int, optional` instead

**Review by skoudoro — COMMENTED** [review]
some additional comments. Thank you again @Sassafrass6.

I am currently checking some vtk features and after your update, it should be good to merge

**Comment by Sassafrass6** [comment]
> > This is necessary for the Cylinder object, which is aligned to the global Y axis in vtk
> 
> Should not it be applied to all actors? I am surprised that only the cylinders are oriented along the Y-axis. I suppose this is the case for all vtksource. Which vtksource is aligned along the X-axis?
> 
> I need to look deeper at this, [when I see this tutorial](https://kitware.github.io/vtk-examples/site/Cxx/GeometricObjects/GeometricObjectsDemo/), they seem to have all the same orientations.

In this example that you have linked, the arrow and cone are aligned along x. The cylinder is aligned along y, and the disk is aligned such that the normal is along z.
The camera is oriented the same way as FURY defaults to, with the x-y plane on parallel with the monitor and the z axis out of the monitor.
The other actors implemented within FURY are all correctly aligned on x, as far as I can tell.

I'm making the format changes now.

-Frank

File: fury/actor.py
Comment (Sassafrass6) [reply fury/actor.py:1614]: This appears to be from an early commit, before I addressed Guaje's comments. In the more recent commits, values for src and rotate are appropriately set to None if faces is not None.

Comment (skoudoro) [reply fury/actor.py:1614]: Can you point which commit? 

I do not find it on master as you can see [here](https://github.com/fury-gl/fury/blob/master/fury/actor.py#L1612)

Comment (skoudoro) [reply fury/actor.py:1614]: Also, can you revert it? thank you

Comment (Sassafrass6) [reply fury/actor.py:1614]: I’m looking on the glyph_orientation branch of my GitHub, [`here`](https://github.com/Sassafrass6/fury/blob/glyph_orientation/fury/actor.py#L1612). 

The link you’ve posted is from fury_gl:master, right?
There is no orientation argument, no disk function, etc. 

What are you referencing, when you ask me to “revert it”?

Comment (skoudoro) [reply fury/actor.py:1614]: Yes, this is `fury-gl:master`. If I merge your PR as it is, this modification will be applied which we do not want. 

So either:
- you rebase your PR
- you merge `master` in your branch.
- you just change this part of the code yourself and commit it.

Sorry for not being explicit, this is what I meant by reverting it (keep this feature already in master).

Comment (skoudoro) [reply fury/actor.py:1614]: Ok, I review the history of this, no need to do any changes, I will go ahead and merge it. Thanks @Sassafrass6

**Comment by skoudoro** [comment]
Thank you for this useful update @Sassafrass6

File: fury/actor.py
Comment (Sassafrass6) [reply fury/actor.py:1614]: I see what you mean now. 
15 commits is a lot to review. Thank you for looking through it all.

**Pull request merged**
